module/VuFind/src/VuFind/RateLimiter/RateLimiterManagerFactory.php

```diff
@@ -0,0 +1,188 @@
+{
+    /**
+     * Service locator
+     *
+     * @var ContainerInterface
+     */
+    protected $serviceLocator;
+
+    /**
+     * Create an object
+     *
+     * @param ContainerInterface $container     Service manager
+     * @param string             $requestedName Service being created
+     * @param null|array         $options       Extra options (optional)
+     *
+     * @return object
+     *
+     * @throws ServiceNotFoundException if unable to resolve the service.
+     * @throws ServiceNotCreatedException if an exception is raised when
+     * creating a service.
+     * @throws ContainerException&\Throwable if any other error occurs
+     */
+    public function __invoke(
+        ContainerInterface $container,
+        $requestedName,
+        array $options = null
+    ) {
+        if (!empty($options)) {
+            throw new \Exception('Unexpected options sent to factory.');
+        }
+
+        $this->serviceLocator = $container;
+
+        $yamlReader = $container->get(\VuFind\Config\YamlReader::class);
+        $config = $yamlReader->get('RateLimiter.yaml');
+
+        $authManager = $container->get(\VuFind\Auth\Manager::class);
+        $request = $container->get('Request');
+
+        return new $requestedName(
+            $config,
+            $request->getServer('REMOTE_ADDR'),
+            $authManager->getUserObject()?->getId(),
+            Closure::fromCallable([$this, 'getRateLimiter']),
+            $container->get(\VuFind\Net\IpAddressUtils::class)
+        );
+    }
+
+    /**
+     * Get rate limiter
+     *
+     * @param array   $config   Rate limiter configuration
+     * @param string  $policyId Policy ID
+     * @param string  $clientIp Client's IP address
+     * @param ?string $userId   User ID or null if not logged in
+     *
+     * @return LimiterInterface
+     */
+    protected function getRateLimiter(
+        array $config,
+        string $policyId,
+        string $clientIp,
+        ?string $userId
+    ): LimiterInterface {
+        $policy = $config['Policies'][$policyId] ?? [];
+        $rateLimiterConfig = $policy['rateLimiterSettings'] ?? [];
+        $rateLimiterConfig['id'] = $policyId;
+        if (null !== $userId && !($policy['preferIPAddress'] ?? false)) {
+            $clientId = "u:$userId";
+        } else {
+            $clientId = "ip:$clientIp";
+        }
+        $factory = new RateLimiterFactory($rateLimiterConfig, $this->createCache($config));
+        return $factory->create($clientId);
+    }
+
+    /**
+     * Create cache for the rate limiter
+     *
+     * @param array $config Rate limiter configuration
+     *
+     * @return ?StorageInterface
+     */
+    protected function createCache(array $config): StorageInterface
+    {
+        if (empty($config['Storage']['namespace'])) {
+            $config['Storage']['namespace'] = 'RateLimiter';
+        }
+        $storageConfig = $config['Storage'] ?? [];
```

Comment:
Because of the namespace check above, I don't think it's possible for `$config['Storage']` to be unset here. Would it make sense to add the namespace to `$storageConfig` instead of the top-level `$config` and then pass `$storageConfig` to `createRedisCache` below, or is there a reason that `createRedisCache` might need access to the full configuration?

Reply:
It was a bit of a mess, should be better now.